packages/core/src/storages/request_queue_v2.ts

```diff
@@ -154,13 +150,80 @@ export class RequestQueue extends RequestProvider {
         //    will not put the request again to queueHeadDict.
         if (request.handledAt) {
             this.log.debug('Request fetched from the beginning of queue was already handled', { nextRequestId });
-            this.recentlyHandledRequestsCache.add(nextRequestId, true);
             return null;
         }
 
         return request;
     }
 
+    /**
+     * @inheritDoc
+     */
+    override async isFinished(): Promise<boolean> {
+        if (this.inProgressRequestBatches.length > 0) {
+            return false;
+        }
+
+        if (this.queueHeadIds.length() > 0) {
+            return false;
+        }
+
+        await this.ensureHeadIsNonEmpty();
+
+        if (this.queueHeadIds.length() > 0) {
+            return false;
+        }
+
+        if (this.queueHasLockedRequests !== undefined) {
+            return !this.queueHasLockedRequests;
```

Comment:
Can we add at least a debug or info log if there are still locked requests? It means that there are locked requests by another client.